tests/periph/timer/main.c

```diff
@@ -51,11 +56,36 @@
  * e.g. when the timer was about to tick anyway */
 #define MINIMUM_TICKS       2
 
+#ifndef TEST_ITERATIONS
+#define TEST_ITERATIONS     (10000ul)
+#endif
+
 static uint8_t fired;
 static uint32_t sw_count;
 static uint32_t timeouts[TIMER_CHANNEL_NUMOF];
 static unsigned args[TIMER_CHANNEL_NUMOF];
 
+typedef struct {
+    unsigned long counter;
+    tim_t dev;
+    mutex_t mtx;
+} test_ctx_t;
+
+static void cb_incr(void *arg, int chan)
+{
+    (void)chan;
+    test_ctx_t *ctx = arg;
+
+    ctx->counter++;
+    if (ctx->counter < TEST_ITERATIONS) {
+        /* Rescheduling the timer like this will trigger a bug in the lptmr
+         * implementation in Kinetis */
+        timer_set(ctx->dev, chan, 20000u);
+        timer_set(ctx->dev, chan, 0);
+    }
+    mutex_unlock(&ctx->mtx);
```

Comment:
Can someone explain to me the purpose of the mutex? It is locked in `test_timer_timeout` and unlocked in the timer callback?

Reply:
That's a common pattern in RIOT for a thread to wait for an IRQ. Unlike a POSIX mutex, a RIOT mutex can be unlocked from a different context (and even from IRQ context).

The idea is that locking an already locked mutex will block. Unlocking that mutex from the ISR will unblock the thread.